# API mini project

This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [ ]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = XXXXXXXXXXXXXX # removed for submission

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [49]:
# First, import the relevant modules
import requests
import collections
import json

In [26]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
r =requests.get("https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2018-09-14&end_date=2018-09-14&api_key=qSych4d1hqEA5JcEA5dB")

In [28]:
print(r.json())

{'dataset': {'id': 10095370, 'dataset_code': 'AFX_X', 'database_code': 'FSE', 'name': 'Carl Zeiss Meditec (AFX_X)', 'description': 'Stock Prices for Carl Zeiss Meditec (AFX) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704', 'refreshed_at': '2018-09-14T22:30:30.323Z', 'newest_available_date': '2018-09-14', 'oldest_available_date': '2000-06-07', 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'frequency': 'daily', 'type': 'Time Series', 'premium': False, 'limit': None, 'transform': None, 'column_index': None, 'start_date': '2018-09-14', 'end_date': '2018-09-14', 'data': [['2018-09-14', None, 76.25, 75.6, 75.95, None, 97874.0, 7434019.0, None, None, None]], 'collapse': None, 'order': None, 'database_id': 6129}}


In [30]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
r.json()

{'dataset': {'collapse': None,
  'column_index': None,
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'data': [['2018-09-14',
    None,
    76.25,
    75.6,
    75.95,
    None,
    97874.0,
    7434019.0,
    None,
    None,
    None]],
  'database_code': 'FSE',
  'database_id': 6129,
  'dataset_code': 'AFX_X',
  'description': 'Stock Prices for Carl Zeiss Meditec (AFX) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'end_date': '2018-09-14',
  'frequency': 'daily',
  'id': 10095370,
  'limit': None,
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'newest_available_date': '2018-09-14',
  'oldest_available_date': '2000-06-07',
  'order': None,
  'premium': False,
  'refreshed_at': '2018-09-14T22:30:30.323Z',
  'start_date': '2018-09-14',
  'transform': None,
  'type': 'Time Series'}}

In [31]:
r.json().keys()

dict_keys(['dataset'])

In [33]:
r.json()['dataset'].keys()

dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])

In [35]:
# Count the number of keys
len(r.json()['dataset'].keys())

21

In [36]:
r.json()['dataset'].values()

dict_values([10095370, 'AFX_X', 'FSE', 'Carl Zeiss Meditec (AFX_X)', 'Stock Prices for Carl Zeiss Meditec (AFX) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704', '2018-09-14T22:30:30.323Z', '2018-09-14', '2000-06-07', ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'daily', 'Time Series', False, None, None, None, '2018-09-14', '2018-09-14', [['2018-09-14', None, 76.25, 75.6, 75.95, None, 97874.0, 7434019.0, None, None, None]], None, None, 6129])

In [44]:
# Count types of values
collections.Counter([type(item) for item in r.json()['dataset'].values()])

Counter({NoneType: 5, bool: 1, int: 2, list: 2, str: 11})

__My note on how nested the data is, as well as the overall structure:__

- The outmost data structure is a dictionary with a unique key 'dataset', whose corresponding value is again a dictionary with 21 keys. 
- The values of those 21 keys are string, integer, None, Boolean and list types. The only keys with a list of values are 'column_names' and 'data'. 
- The value of the key __'column_names'__ is one list with column names in string types. 
- The __'data'__ key has a list of lists as a value. Each inner list has a data for each day and it seems the data corresponds to the column names in 'column_names'.

These are tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

### 1. Collect data

In [45]:
r_2017 =requests.get("https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=qSych4d1hqEA5JcEA5dB")

### 2. Convert to Python dictionary 

In [55]:
# convert json object to python dictionary
data_dict = json.loads(json.dumps(r_2017.json()))

This works, but later I found r_2017.json() is already a python dictionary.

In [59]:
type(r_2017.json())

dict

In [63]:
r_2017.json() == json.loads(json.dumps(r_2017.json()))

True

They are indeed the same data. I will use r_2017.json() since it is more simple.

In [65]:
data_dict = r_2017.json()

### 3. Highest and lowest opening prices

In [66]:
# Check column names
data_dict['dataset']['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [162]:
data_dict['dataset']['column_names'].index('Open')

1

The opening prices are the second (index=1) item in the list 'column_names', so I will use the second item from 'data'

In [70]:
open_prices = [daily_data[1] for daily_data in data_dict['dataset']['data']]

In [214]:
# Count missing values
sum([1 for price in open_prices if price is None])

3

There are 3 missing values. I will remove them since there are not too many of them.

In [105]:
open_prices_no_NA = [daily_data[1] for daily_data in data_dict['dataset']['data'] if daily_data[1] is not None]

In [97]:
print( "Highest opening price:", max(open_prices_no_NA))
print( "Lowest opening price:", min(open_prices_no_NA))

Highest opening price: 53.11
Lowest opening price: 34.0


- __The highest opening price is 53.11__
- __The lowest opening price is 34.0__

### 4. Largest change in any one day (based on High and Low price)

In [163]:
data_dict['dataset']['column_names'].index('High'), data_dict['dataset']['column_names'].index('Low')

(2, 3)

The high and low prices are the third (index=2) and fourth (index=3) items in the data, respectively.

In [215]:
# Count missing high prices
sum([1 for daily_data in data_dict['dataset']['data'] if daily_data[2] is None])

0

In [216]:
# Count missing low prices
sum([1 for daily_data in data_dict['dataset']['data'] if daily_data[3] is None])

0

There are no missing value for high or low prices

In [217]:
change_daily = [daily_data[2] - daily_data[3] for daily_data in data_dict['dataset']['data']]

In [221]:
print("%.2f" % max(change_daily))

2.81


__The largest change in any one day is 2.81__

### 5. Largest change between any two days (based on Closing Price)

This problem is a little unclear. I wonder if 'any two days" here mean (1) any two consecutive days or they mean (2) any two different days that can be several days apart. I will get solutions for both cases.

(1) Largest change between any two __consecutive__ days 

In [164]:
data_dict['dataset']['column_names'].index('Close')

4

The closing price is the fifth (index=4) entry of data.

In [222]:
# Count missing closing prices
sum([1 for daily_data in data_dict['dataset']['data'] if daily_data[4] is None])

0

There is no missing closing value.

In [224]:
closing_prices = [daily_data[4] for daily_data in data_dict['dataset']['data']]

In [225]:
# Possible first day closing prices (the last day dropped)
closing_day1 = closing_prices[:-1] 
# Possible second day closing prices (the first day dropped)
closing_day2 = closing_prices[1:]

In [226]:
print("Largest increase: %.2f" % max([day2 - day1 for day1, day2 in zip(closing_day1, closing_day2)]))

Largest increase: 2.56


In [157]:
print("Largest drop: %.2f" % min([day2 - day1 for day1, day2 in zip(closing_day1, closing_day2)]))

Largest drop: -1.72


(2) Largest change between any two different days that can be several days apart

In [227]:
print("Largest change: %.2f" % (max(closing_prices)-min(closing_prices)))

Largest change: 19.03


- __The largest change between any two consecutive days is 2.56.__
- __The largest change between any two days (not necessarily consecutive) is 19.03.__


### 6. Average daily trading volume during 2017

In [165]:
data_dict['dataset']['column_names'].index('Traded Volume')

6

The trading volumn is the seventh (index=6) item of data.

In [230]:
# Count missing trading volumes
sum([1 for daily_data in data_dict['dataset']['data'] if daily_data[6] is None])

0

There is no missing trading volume.

In [231]:
trading_volume = [daily_data[6] for daily_data in data_dict['dataset']['data']]

In [232]:
print("Average daily trading volumn: %.2f" %(sum(trading_volume)/len(trading_volume)))

Average daily trading volumn: 89124.34


__The average daily trading volumn is 89124.34__

### 7. (Optional) Median trading volume during 2017

In [209]:
# make a function that calculate median from a Python list
def median(my_list):
    size = len(my_list)
    sorted_list = sorted(my_list)
    if size%2 == 1:
        return sorted_list[size//2]
    else:
        return (sorted_list[size//2-1]+sorted_list[size//2])/2

In [210]:
median(trading_volume)

76286.0

In [211]:
# Double check with numpy (numpy was used just for checking)
import numpy as np
np.median(np.array(trading_volume))

76286.0

__The median trading volumn is 76286.0__